### ISSUE TITLE
Usability issue: complicated to separate between multiplication/pow and argument unpacking

### ISSUE THREAD

**Issue by boxed** [davidhalter/parso#54]
The * in `foo * bar` produces produces an operator AST node with value `*`, same as the * in `foo(*bar)` even though these are radically different things. The same goes for `**`. It would be much nicer if these were differentiated in the AST clearly.

This was found when working on mutmut.

**Comment by davidhalter** [comment]
I don't really understand, because at least for `*` we have different nodes. For multiplication it's `term`, for functions it's Jedi's own `param` and for function calls it's `argument` (Python 3.5+).

Also `**` is very clearly differentiated, because it's a `power` node. Before Python 3.4+ this is a little bit less clear, because `atom_expr` and `power` are mixed.

**Comment by boxed** [comment]
For function definitions it's not the case:

```
In [11]: parse('def foo(a, *, b): pass').children[0].children[2].children[2].type                                                          
Out[11]: 'operator'
```

For `**` it's a bit annoying too:

```
In [15]: parse('def foo(a, **foo): pass').children[0].children[2].children[2].children[0]                                                  
Out[15]: <Operator: **>

In [16]: parse('a ** b').children[0].children[1]                                                                                           
Out[16]: <Operator: **>
```

At the very least there should be a difference between an "operator with the name **" and "a unary operator with the name **". They just happen to share some syntax but are totally different things.

I hope this all makes sense.. obviously I was waaaaay too vague in my initial description. Sorry about that!

**Comment by davidhalter** [comment]
Hmm ok. I get what you're saying. You're speaking about the leaf itself. My thought was that you could check `leaf.parent == 'power'` and you can match it pretty well. However I agree that it's more complicated than it could be.

I'll keep it open for now and maybe will find a good solution in the future. But I'm not sure something will change, because the information is there and it would break backwards compatibility.

**Comment by boxed** [comment]
Exactly. Matching on just the leaf would make my mutation code in mutmut much much simpler. Sometimes I can do this, but sometimes not. The code to match the surrounding AST nodes is currently very ugly, but I'm working on a general matching system to improve this. Here's a test case for an experimental branch: https://github.com/boxed/mutmut/pull/62/files#diff-5bf95ee98ceeee163338bb2ae2366b60R30

I think a matching system like this would be a good addition to parso eventually since it could make it super easy to write refactorings, search, etc.

> and it would break backwards compatibility

Yea, this is always a problem. One could handle this by adding a `subtype` property though. It's a bit awkward compared to having its own type, but way better than looking at a lot of the context. When I look at the surrounding context I just stop the debugger and look what I have and basically treat it as a black box. It's not at all obvious that any of that code is correct :P

**Comment by boxed** [comment]
> Is x.parent.type == 'term' really much harder than x.sub_type == 'multiplication'? Obviously it's a bit better, but I don't see it as a big improvement.

I disagree. The problem is that how do you know if you need to look at the parent type for a node or not? Do you also need to look at parent.parent? parent.children[0]? Etc etc. I do actually have a lot of code like that in mutmut that is pure trial and horror. Pretty sure a lot of it is wrong subtly :P

> BTW since I read your article, I feel like Python ~3.5+ importlib is pretty good if you want to work around the import system. It's not perfect, but it works pretty well. It obviously sucks in Python 2.

Which article? I feel like I didn't follow the change of topic here....

**Comment by boxed** [comment]
It wasn't obvious to me that I needed to look there. I admit I am terrible at reading docs so maybe it says so there and I just didn't look. 

But to me it seems the docs and parso don't agree fully. Looking at the arguments definition it's not "the operator * followed by.." But just "the character * followed by...." So reading the grammar seems to indicate one should not get an operator node there. *shrug*

**Comment by boxed** [comment]
The rest of the paragraph was pretty exact I think. Pardo treats the * in `def foo(*args)` as an operator, multiplication specifically. The python grammar spec does not. It's just a * character.
